kep/9-capacity-scheduling-for-ml-dl-workloads-based-on-scheduler-framework/README.md

```diff
@@ -70,24 +73,24 @@ updates.
+consumption may be much smaller than the limit.
 
-1. The quota resource usage is aggregated based on the resource configurations (e.g., Pod cpu/mem requests specified in the Pod spec). Although this mechanism can guarantee that the actual resource consumption will never exceed the ResourceQuota limit, it might lead to low resource utilization had the actual resource consumption been much smaller than the limit. i.e., lead to internal resource fragmentation.
+Due to above limitation, the batch workloads (ML/DL) can't run in a Kubernetes cluster as efficiently as they do in other container orchestration platforms such as Yarn. In order to overcome above limitations, we introduce an "ElasticQuota" concept used in [Yarn capacity scheduler](https://www.netjstech.com/2018/04/capacity-scheduler-in-yarn-hadoop.html) in Kubernetes. Basically, the "ElasticQuota" has the notions of "max" and "min", where the "min” is the minimum resources that are needed to ensure the basic functionality/performance of the consumers and the "max" specifies the upper bound of the resource consumption of the consumers. By introducing "min" and "max", capacity scheduling can be optimized as follows:
 
-2. If we use ResourceQuota to strictly divide the resources of the cluster among all users to prevent clusters from running out of resources, it might also lead to low resource utilization because some users may have quota resources not being used.
+1. The resource buffer between "min" and "max" can help to tolerate runtime failures. For example, if a Pod that consumes the entire "min" fails to run, new Pods can still be created if the "max" has not been reached. When using Kubernetes resource quota, once the Pod that consumes the entire quota is created,
+no other Pods can be created - even if the Pod failed to be scheduled, or failed on running (error
+when pulling image, etc.).
 
-Due to above limitations, the batch workloads (ML/DL) can't run in a Kubernetes cluster as efficiently as they do in other container orchestration platforms such as Yarn. In order to overcome above limitations, we introduce an "Queue" concept used in Yarn capacity scheduler [1] in Kubernetes. Basically, the "Queue" has the notions of "max" and "min", where the "min” is the minimum resources that are needed to ensure the basic functionality/performance of the consumers and the "max" specifies the upper bound of the resource consumption of the consumers. By introducing "min" and "max",  Pod scheduling allows the following optimizations:
+2. Improve overall resource utilization by allowing the workloads from one quota to "borrow" unused reserved "min" resources from other quotas. A quota's unused "min" resource can be used by other users, under the condition that there is a mechanism to guarantee the "victim" user can consume his "min" resource whenever he needs. Typically, this is done by implementing preemption.
```

Comment:
Improve overall resource utilization by allowing the workloads... ->
Allow the workloads...

his -> its, he -> it
~~Typically, this is done by implementing preemption.~~

Reply:
Another question is: when to start the preemption? It seems that if the system resource is efficient, all elastic quotas can be fully used up to its "max". So doe that mean preemption can only happen when system resource has been used up globally?

Reply:
Yes. When 
1. there are free resources in the cluster
2.the allocated of quota has not reached max

The preemption will happen If one of the two items is not satisfied.